File: algorithm/leetcodeCpp/populating-next-right-pointers-in-each-node-ii.cpp
// 117. Populating Next Right Pointers in Each Node II
// Difficulty: Hard

// Follow up for problem "Populating Next Right Pointers in Each Node".

// What if the given tree could be any binary tree? Would your previous solution still work?

// Note:

// You may only use constant extra space.
// For example,
// Given the following binary tree,
//          1
//        /  \
//       2    3
//      / \    \
//     4   5    7
// After calling your function, the tree should look like:
//          1 -> NULL
//        /  \
//       2 -> 3 -> NULL
//      / \    \
//     4-> 5 -> 7 -> NULL

// Time:  O(n)
// Space: O(1)

class Solution {
public:
    void connect(TreeLinkNode *root) {
        while (root) {
            TreeLinkNode* next = nullptr; // the first node of next level
            for (TreeLinkNode* prev = nullptr; root; root = root->next) {
                if (!next) {
                    next = root->left ? root->left : root->right;
                }
                if (root->left) {
                    if (prev) {
                        prev->next = root->left;
                    }
                    prev = root->left;
                }
                if (root->right) {
                    if (prev) {
                        prev->next = root->right;
                    }
                    prev = root->right;
                }
            }
            root = next; // turn to next level
        }
    }
};

File: zZ/java/populating-next-right-pointers-in-each-node-ii.java
/**
 * Populating Next Right Pointers in Each Node II
 * 
 * Given a binary tree
 * 
 * struct TreeLinkNode {
 *   TreeLinkNode *left;
 *   TreeLinkNode *right;
 *   TreeLinkNode *next;
 * }
 * Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to NULL.
 * 
 * Initially, all next pointers are set to NULL.
 * 
 * Note:
 * 
 * You may only use constant extra space.
 * Recursive approach is fine, implicit stack space does not count as extra space for this problem.
 * Example:
 * 
 * Given the following binary tree,
 * 
 *      1
 *    /  \
 *   2    3
 *  / \    \
 * 4   5    7
 * After calling your function, the tree should look like:
 * 
 *      1 -> NULL
 *    /  \
 *   2 -> 3 -> NULL
 *  / \    \
 * 4-> 5 -> 7 -> NULL
 */

public class Solution {
    public void connect(TreeLinkNode root) {
        for (TreeLinkNode next = null; root != null; root = next) {
            for (TreeLinkNode curr = null; root = root.next; root = root.next) {
                if (next == null) {
                    next = root.left ? root.left : root.right;
                }

                if (root.left != null) {
                    if (curr != null) {
                        curr.next = root.left;
                    }
                    curr = root.left;
                }

                if (root.right != null) {
                    if (curr != null) {
                        curr.next = root.right;
                    }
                    curr = root.right;
                }
            }
        }
    }
}

public class TreeLinkNode {
    int val;
    TreeLinkNode left, right, next;
    TreeLinkNode(int x) { val = x; }
}